### ISSUE TITLE
reduce allocations when using New()

### ISSUE THREAD

**Issue by ongardie-ebay** [cespare/xxhash#13]
Hey Caleb,

Thanks for the library, and we should catch up soon. Anyway, we were benchmarking the following sequence:
```go
h := cespare.New()
h.Write(p)
h.Write(o)
h.Sum64()
```
and found that this causes 1 memory allocation. We then changed `New()` from returning  a `hash.Hash64` to returning a `*xxh`. This removes the memory allocation and saves about 25ns per op on my MacBook Pro. Would you be open to that? It probably makes sense to make the `xxh` struct public at that point.

Second and pending the above, one of our key use cases involves hashing a uint64 followed by a string. It'd be useful to have a `WriteString` on `*xxh` that would avoid a memory allocation (analogous to `Sum64String`).

Thanks,
Diego

**Comment by cespare** [comment]
Hey Diego, nice to hear from you :)

This change seems pretty reasonable, and it's probably how I would write the API if I were doing it today.(This issue identifies the two main problems with returning interfaces rather than concrete types: it forces allocations and it makes it hard to add new methods in the future.) I did it like this initially to mimic the stdlib functions like [sha256.New](https://golang.org/pkg/crypto/sha256/#New).

WriteString certainly makes sense as a follow-on as well.

If you'd like to send the change, that'd be great!

`xxh` isn't a great name to begin with; since we need to export it, let's call it `Digest`.

(It's a backward-incompatible change, so we'll need to bump the version to 2.0.0 afterwards.)

**Comment by ongardie-ebay** [comment]
Cool, I'll prep a PR (but it might take me a few days). I see why we need to bump the minor version for each of these changes, but what makes it a backwards-incompatible change? Since `Digest` will still implement `hash.Hash64`, I had assumed that existing callers would continue to work. Can you give a code example that would break?

**Comment by cespare** [comment]
```
var myHashFunc func() hash.Hash64
myHashFunc = xxhash.New
```

**Comment by ongardie-ebay** [comment]
Gotcha, thanks.

**Comment by ongardie-ebay** [comment]
I made some progress on this today (but I need legal's permission before I can post it). I'm not sure what I've done differently from the other day, but I'm finding that I've removed all the allocations in the purego build only -- one remains for the asm build.

I believe the problem is with:
```go
func writeBlocks(x *Digest, b []byte) []byte
```
This function isn't marked as `go:noescape`, so the compiler can't tell that the `Digest` there doesn't escape out to the heap.

The docs on `go:noescape` say:
> [the function] does not allow any of the pointers passed as arguments to escape into the heap or into the values returned from the function

Since `writeBlocks` returns a subslice of its second argument, it seems to me like we're not allowed to simply mark `writeBlocks` as `go:noescape`. (But I'm not certain. I did try this, and it did result in 0 allocations, but it's probably not safe.)

One option would be to have `writeBlocks` return an integer representing the number of bytes it wrote. Then the caller could "increment" its slice accordingly. @cespare, can you make that change? I'm too scared to touch the assembly. It's also fine if you want to wait until seeing my patch set, just let me know.

---

Here are my current benchmark results:

    $ go test -benchtime 10s -benchmem -bench '/[xO].*n=100_B'
    goos: darwin
    goarch: amd64
    pkg: github.com/cespare/xxhash
    BenchmarkHashes/xxhash,direct,bytes,n=100_B-4           1000000000              16.7 ns/op      5970.94 MB/s           0 B/op          0 allocs/op
    BenchmarkHashes/xxhash,direct,string,n=100_B-4          1000000000              18.9 ns/op      5283.42 MB/s           0 B/op          0 allocs/op
    BenchmarkHashes/xxhash,digest,bytes,n=100_B-4           300000000               51.1 ns/op      1956.42 MB/s          80 B/op          1 allocs/op
    BenchmarkHashes/xxhash,digest,string,n=100_B-4          300000000               54.6 ns/op      1830.61 MB/s          80 B/op          1 allocs/op
    BenchmarkHashes/OneOfOne,direct,bytes,n=100_B-4         1000000000              22.9 ns/op      4372.87 MB/s           0 B/op          0 allocs/op
    BenchmarkHashes/OneOfOne,direct,string,n=100_B-4        500000000               24.6 ns/op      4069.27 MB/s           0 B/op          0 allocs/op
    BenchmarkHashes/OneOfOne,digest,bytes,n=100_B-4         500000000               38.8 ns/op      2577.23 MB/s           0 B/op          0 allocs/op
    BenchmarkHashes/OneOfOne,digest,string,n=100_B-4        300000000               44.8 ns/op      2231.34 MB/s           0 B/op          0 allocs/op
    PASS
    ok      github.com/cespare/xxhash       162.815s
    $ go test -tags purego -benchtime 10s -benchmem -bench '/x.*n=100_B'
    goos: darwin
    goarch: amd64
    pkg: github.com/cespare/xxhash
    BenchmarkHashes/xxhash,direct,bytes,n=100_B-4           500000000               23.3 ns/op      4296.41 MB/s           0 B/op          0 allocs/op
    BenchmarkHashes/xxhash,direct,string,n=100_B-4          1000000000              20.1 ns/op      4975.62 MB/s           0 B/op          0 allocs/op
    BenchmarkHashes/xxhash,digest,bytes,n=100_B-4           500000000               29.1 ns/op      3434.35 MB/s           0 B/op          0 allocs/op
    BenchmarkHashes/xxhash,digest,string,n=100_B-4          500000000               32.2 ns/op      3107.97 MB/s           0 B/op          0 allocs/op
    PASS
    ok      github.com/cespare/xxhash       75.974s

**Comment by cespare** [comment]
Your suggestion seems good to me. I took a quick pass at it in #14 and it all seems to work (but I need to clean that up a bit). A review of that would be appreciated.

When you send PR(s), set the base to that branch (digest) and I'll merge them all at the same time.

Side note: just to check, in your use case, can you not reuse the digest (with `Reset`)? That's a typical pattern for avoiding allocations even if the digest itself does escape to the heap. That is, instead of

```
	for i := 0; i < b.N; i++ {
		h := xxhash.New()
		h.Write(input)
		_ = h.Sum64()
	}
```

you can do

```
	h := xxhash.New()
	for i := 0; i < b.N; i++ {
		h.Reset()
		h.Write(input)
		_ = h.Sum64()
	}
```

(It's reasonable that the answer would be "no" -- for a lightweight hash like xxhash that has a relatively small amount of internal state it is certainly useful in some scenarios to cheaply create and discard digests on the fly. Just want to confirm that you considered it.)

**Comment by cespare** [comment]
BTW for benchmark results you include in a PR/commit message, please split the bench time across several smaller runs (`-benchtime 3s -count 7`, say) and summarize the results with [benchstat](https://godoc.org/golang.org/x/perf/cmd/benchstat).

**Comment by ongardie-ebay** [comment]
Awesome, thanks. We commonly compute the hashes to figure out where to send requests, and in that request context, we wouldn't necessarily get to reuse a Digest. It's also nice to not contort our code around reusing a Digest that's probably cheap enough to drop on the stack anyway.

**Comment by cespare** [comment]
OK, I released v2.0.0 which includes your changes. Thanks for the contribution!

**Comment by ongardie-ebay** [comment]
Sweet. Thanks @cespare!
